### DISCUSSION TITLE
Improve 204 (No content) handling and polymorphism

### DISCUSSION THREAD

**Discussion by sky-franciscoaleixo** [slackhq/EitherNet#79] [Ideas] [answered]
### Description
There has been a PR (https://github.com/slackhq/EitherNet/pull/63) that introduces the capability of returning ApiResult<Unit, ...> when a 204 occurs. This is generally fine if we expect that  that is the only successful behavior.

But what if an endpoint can return a 200 with a specific body response, but can also return a 204 with no body? I'd like to open discussion and maybe ask for some guidance regarding this. I've also seen the README which kind of talks about 200 polymorphism but it again seems to only relate with errors. There's a few options to resolve this:
- Change `ApiResult<out T : Any, out E : Any>`, so that at least T can be nullable. This would allow nullable responses, and fix this specific issue, but there may be reasons as to why you might not want to do this.
- Allow some way to pass HTTP status code through Retrofit annotations in success responses. This way we could implement custom behavior in custom `Converter.Factory`s. I tried this, and this is already possible in case of an error, but not in the case of success AFAIK.
- I could try catch NPE, but that's not really a good solution
- I could use Response<MyDto?> and try to convert manually to ApiResult, but then I'd have to reimplement the whole CallAdapter logic.

Is there any other option? Thank you.

### Requirements (place an `x` in each of the `[ ]`)
* [x] I've read and understood the [Contributing guidelines](../CONTRIBUTING.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

**Comment by ZacSweers** [comment]
Retrofit handles this by just coercing to `Unit`, no?

**Reply by sky-franciscoaleixo** [reply]
Sorry for the delay (was on vacation), I think Retrofit does coerce a 'no body' response to `Unit` if you specify a `Unit` response, but doing `Response<MyDto?>` works as it does return null when there's no body.

If I do `ApiResult<MyDto, Throwable>` and I get a no body response (on 204), I will get a NPE though, and I can't use a nullable type on ApiResult, hence the question.

**Reply by ZacSweers** [reply]
I see, I'm not sure what to do about that situation. What do other null-averse call adapters do for that situation, such as RxJava 2/3?

**Reply by sky-franciscoaleixo** [reply]
Well I haven't used RxJava in a long while, but interestingly you had a PR for their CallAdapter https://github.com/square/retrofit/pull/3003 related with that 😆 

Regardless, their stance seems to be this https://github.com/square/retrofit/issues/2242#issuecomment-300850574

Side note, it seems like Retrofit2 also foces null body when it receives a 204/205.

**Reply by ZacSweers** [reply]
Heh, good find. I suspect the `Optional` approach there is your best bet

**Reply by sky-franciscoaleixo** [reply]
Thanks for the response! I guess my last question on this is, is there any technical reason we can't have nullable type on ApiResult? Is it just a philosophical stance, or is there any other reason?

**Reply by ZacSweers** [reply]
It's not possible in the type system to only allow this for 204 responses. If we allowed it here we'd allow it everywhere, and the whole point of this library is to require stronger types. Not only this, but there's no way at the Java level to inspect the type for nullability without metadata parsing since it's a kotlin construct.

**Reply by sky-franciscoaleixo** [reply]
Late answer from me, but that's fair, thanks!

**Comment by sky-franciscoaleixo** [comment]
I'm going to close this discussion and suggest if anyone has this same problem to follow the Optional approach suggested above 👍

**Comment by sky-franciscoaleixo** [comment]
I'm reopening this as using `Optional<>` won't work. This is because as mentioned above Retrofit forces a null response, but this also means that it won't pass through any Converter at all, so using an `OptionalConverter` won't ever work.

The only options I see are either intercepting these 204/205 requests and forcing some other response, or hammering some logic on ApiResultCallAdapter to support Optional there which doesn't make any sense. If I end up finding an acceptable workaround I'll post here.

**Reply by ZacSweers** [reply]
Is this really an EitherNet issue at this point? Seems more a limitation of Retrofit

**Reply by sky-franciscoaleixo** [reply, marked as answer]
I'd say at this point it is mostly a limitation of Retrofit, I just left this as other people could find themselves in the same situation. I'll leave the solution that works ok below with an expected return of `ApiResult<Optional<T>, Throwable>` and a specific annotation `@OptionalNoContent`, although I don't really like it.

```kotlin
internal class NoContentInterceptor : Interceptor {
    private val noContentHttpStatusCodes = arrayOf(204, 205)

    override fun intercept(chain: Interceptor.Chain): Response {
        val request = chain.request()
        val response = chain.proceed(request)
        
        if (request.getAnnotation<OptionalNoContent>() == null ||
            !response.isSuccessful ||
            !noContentHttpStatusCodes.contains(response.code)
        ) {
            return response
        }

        return response
            .newBuilder()
            .code(200)
            .body("".toResponseBody("text/plain".toMediaType()))
            .build()
    }
}
```
and

```kotlin
internal object NoContentConverterFactory : Converter.Factory() {
    override fun responseBodyConverter(
        type: Type,
        annotations: Array<out Annotation>,
        retrofit: Retrofit,
    ): Converter<ResponseBody, *>? {
        if (getRawType(type) != Optional::class.java || annotations.find { annotation -> annotation is OptionalNoContent } == null) {
            return null
        } else {
            val innerType = getParameterUpperBound(0, type as ParameterizedType)
            val delegate = retrofit.responseBodyConverter<Any?>(innerType, annotations)
            return NoContentConverter(delegate)
        }
    }

    private class NoContentConverter<T : Any>(
        private val delegate: Converter<ResponseBody, T>? = null,
    ) : Converter<ResponseBody, Optional<T>> {

        override fun convert(value: ResponseBody): Optional<T> {
            if (value.contentLength() <= 0L) {
                return Optional.empty()
            }
            return Optional.ofNullable(delegate?.convert(value))
        }
    }
```
